kernel/kernel-api/src/main/java/io/delta/kernel/internal/util/ColumnMapping.java

```diff
@@ -176,4 +179,96 @@ public static String getPhysicalName(StructField field) {
+        String newMappingMode = getColumnMappingMode(newConfig);
+
+        Preconditions.checkArgument(isNewTable ||
+                        allowMappingModeChange(oldMappingMode, newMappingMode),
+                "Changing column mapping mode from '%s' to '%s' is not supported",
+                oldMappingMode, newMappingMode);
+    }
+
+    private static boolean allowMappingModeChange(String oldMode, String newMode) {
+        // only upgrade from none to name mapping is allowed
+        return oldMode.equals(newMode) ||
+                (COLUMN_MAPPING_MODE_NONE.equals(oldMode) &&
+                        COLUMN_MAPPING_MODE_NAME.equals(newMode));
+    }
+
+    public static boolean isColumnMappingModeEnabled(String columnMappingMode) {
+        return NAME_OR_ID.contains(columnMappingMode);
+    }
+
+    public static Metadata updateColumnMappingMetadata(Metadata metadata,
+                                                       String columnMappingMode,
+                                                       boolean isNewTable) {
+        switch (columnMappingMode) {
+            case COLUMN_MAPPING_MODE_NONE:
+                return metadata;
+            case COLUMN_MAPPING_MODE_ID: // fall through
+            case COLUMN_MAPPING_MODE_NAME:
+                return assignColumnIdAndPhysicalName(metadata, isNewTable);
+            default:
+                throw new UnsupportedOperationException(
+                        "Unsupported column mapping mode: " + columnMappingMode);
+        }
+    }
+
+    private static Metadata assignColumnIdAndPhysicalName(Metadata metadata, boolean isNewTable) {
+        StructType schema = metadata.getSchema();
+        // TODO: read max col id from delta.columnMapping.maxColumnId
+        int maxColumnId = findMaxColumnId(schema);
+        StructType newSchema = new StructType();
+        for (StructField field : schema.fields()) {
+            StructField updatedField = field;
+
+            if (!hasColumnId(field)) {
+                maxColumnId++;
+                updatedField = field.withNewMetadata(FieldMetadata.builder()
+                        .fromMetadata(field.getMetadata())
+                        .putLong(COLUMN_MAPPING_ID_KEY, maxColumnId)
+                        .build());
+            }
+            if (!hasPhysicalName(field)) {
+                // re-use old display names as physical names when a table is updated
+                String physicalName = isNewTable ? "col-" + UUID.randomUUID() : field.getName();
+                updatedField = updatedField.withNewMetadata(FieldMetadata.builder()
+                        .fromMetadata(updatedField.getMetadata())
+                        .putString(COLUMN_MAPPING_PHYSICAL_NAME_KEY, physicalName)
+                        .build());
+            }
+            newSchema = newSchema.add(updatedField);
+        }
+
+        return metadata.withNewSchema(newSchema);
+    }
+
+    static int findMaxColumnId(StructType schema) {
```

Comment:
I initially had it private but made it package-private so that it can be used in unit tests (see `TestColumnMapping.findMaxColumnId`). I'll be adding nested field support in follow-up PR and would like to have proper unit tests for this particular method

Reply:
Sounds good! Can you add ` /** Visible for testing */` above the method, then?

Reply:
done